apps/libsignpost/port_signpost_arduino.cpp

```diff
@@ -0,0 +1,252 @@
+	//Set global i2c slave transmit buffer
+	g_slave_transmit_buf = buf;
+	g_slave_transmit_buf_len = len;
+	return 0;
+}
+
+int port_signpost_mod_out_set() {
+	digitalWrite(ARDUINO_MOD_OUT, HIGH);
+	return 0;
+}
+
+int port_signpost_mod_out_clear() {
+	digitalWrite(ARDUINO_MOD_OUT, LOW);
+	return 0;
+}
+
+int port_signpost_mod_in_read() {
+	return digitalRead(ARDUINO_MOD_IN);
+}
+
+int port_signpost_mod_in_enable_interrupt_falling(port_signpost_callback cb) {
+	//Set global MOD_IN falling edge callback 
+	g_mod_in_callback_falling = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_falling_helper, FALLING);
+	return 0;
+}
+
+int port_signpost_mod_in_enable_interrupt_rising(port_signpost_callback cb) {
+	//Set global MOD_IN rising edge callback 
+	g_mod_in_callback_rising = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_rising_helper, RISING);
+	return 0;
+}
+
+int port_signpost_mod_in_disable_interrupt(void) {
+	detachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN));
+	return 0;
+}
+
+//TODO: Implement version that puts Arduino into a sleep state
+void port_signpost_wait_for(void* wait_on_true) {
+	volatile bool wait_on_true_volatile = false;
+	//Stay in loop until wait_on_true becomes true
+	while (!wait_on_true_volatile) {
+		wait_on_true_volatile = *((bool*) wait_on_true);
+	}
+}
+
+//TODO: Implement version that actually times out
+//		For now it is just a replica of port_signpost_wait_for
+int port_signpost_wait_for_with_timeout(void* wait_on_true, uint32_t ms) {
+	port_signpost_wait_for(wait_on_true);
+	return SB_PORT_SUCCESS;
+}
+
+void port_signpost_delay_ms(unsigned ms) {
+	//For every 1 ms delay, call delayMicroseconds once.
+	//delayMicroseconds does not work for inputs > 16383,
+	//so a loop is used to divide up the duration.
+	for (unsigned i = 0; i < ms; ++i) {
+		delayMicroseconds(1000);
+	}
+}
+
+int port_signpost_debug_led_on() {
+	digitalWrite(DEBUG_LED, HIGH);
+	return 0;
+}
+
+int port_signpost_debug_led_off() {
+	digitalWrite(DEBUG_LED, LOW);
+	return 0;
+}
+
+int port_rng_init() {
+    if (signpost_entropy_init() < 0) return SB_PORT_FAIL;
+    return SB_PORT_SUCCESS;
+}
+
+//Uses pseudo-random number generation 
+//TODO: Connect hardware rng to arduino and poll that device for random numbers
+int port_rng_sync(uint8_t* buf, uint32_t len, uint32_t num) {
+	//Generate pseudo-random seed by reading an analog pin
+	randomSeed(analogRead(2)); 
+	uint32_t i = 0;
+	//Generate pseudo-random numbers from seed
+	for (; i < num && i < len; ++i) {
+		buf[i] = random(0, 256) & 0xFF;
+	}
+	return i;
+}
+
+int port_printf(const char *fmt, ...) {
+	#ifdef ENABLE_ARDUINO_PORT_PRINTF
+		va_list args;
+		va_start(args, fmt);
+		vsnprintf(g_port_printf_buf, _PRINTF_BUFFER_LENGTH_, fmt, args);
+		Serial.print(g_port_printf_buf);
+		va_end(args);
+	#endif
+	return 0;
```

Comment:
Hmm... I think printf family of functions usually return the number of bytes written (not that this will probably matter too much in practice). I think you should return the return value of `vsnprintf` above in the enabled case. 0 in the disabled case is fine.